File: api/handlers/dtos/requests/breed.go
package requests

// Any json body that requires a breed name.
// This is for use with validation.
type BreedNameRequiredJson struct {
	Name string `json:"name" binding:"required"`
}

// Same for Uri.
type BreedIdRequiredUri struct {
	Id uint `uri:"id" binding:"required"`
}

type BreedIdOptionalUri struct {
	Id uint `uri:"id"`
}

type BreedCategoryIdRequiredJson struct {
	// Avoid circular references.
	CategoryId uint `json:"category_id" binding:"required"`
}

type GetBreed struct {
	BreedIdRequiredUri
}

type AddBreed struct {
	BreedNameRequiredJson
	BreedCategoryIdRequiredJson
}

type PutBreedUri struct {
	BreedIdOptionalUri
}

type PatchBreedUri struct {
	BreedIdOptionalUri
}

type PatchBreedBody struct {
	BreedNameRequiredJson
}

File: website/src/lib/dtos/requests/breed.ts
// ...new file...
import { RequestInfo } from './requestInfo';

export interface BreedNameRequiredJson {
  name: string;
}

export interface BreedIdRequiredUri {
  id: number;
}

export interface BreedIdOptionalUri {
  id?: number;
}

export interface BreedCategoryIdRequiredJson {
  category_id: number;
}

/** Request payload to add a breed */
export interface AddBreed {
  name: string;
  category_id: number;
}

/** Classes that implement the request shapes and provide toRequestInfo() */
export class GetBreed implements BreedIdRequiredUri {
  id: number;

  constructor(id: number) {
    this.id = id;
  }

  toRequestInfo() {
    return new RequestInfo({ body: { class: 'GetBreed', id: this.id } });
  }
}

export class PutBreedUri implements BreedIdOptionalUri {
  id?: number;

  constructor(id?: number) {
    this.id = id;
  }

  toRequestInfo() {
    return new RequestInfo({ body: { class: 'PutBreedUri', id: this.id } });
  }
}

export class PatchBreedUri implements BreedIdOptionalUri {
  id?: number;

  constructor(id?: number) {
    this.id = id;
  }

  toRequestInfo() {
    return new RequestInfo({ body: { class: 'PatchBreedUri', id: this.id } });
  }
}

export class PatchBreedBody implements BreedNameRequiredJson {
  name: string;

  constructor(name: string) {
    this.name = name;
  }

  toRequestInfo() {
    return new RequestInfo({ body: { class: 'PatchBreedBody', name: this.name } });
  }
}

export class DeleteBreedUri implements BreedIdRequiredUri {
  id: number;

  constructor(id: number) {
    this.id = id;
  }

  toRequestInfo() {
    return new RequestInfo({ body: { class: 'DeleteBreedUri', id: this.id } });
  }
}

export const breedEndpoints = {
  addBreed: '/breed',
  getBreed: (id: number) => `/breeds/${id}`,
  putBreed: (id?: number) => id != null ? `/breeds/${id}` : '/breeds',
  patchBreed: (id?: number) => id != null ? `/breeds/${id}` : '/breeds',
  deleetBreed: (id: number) => `/breeds/${id}`,
} as const;

